compaction/task_manager_module.cc

```diff
@@ -422,7 +422,15 @@ future<> table_major_keyspace_compaction_task_impl::run() {
 }
 
 future<> cleanup_keyspace_compaction_task_impl::run() {
+    std::vector<table_id> table_ids;
+    std::ranges::transform(_table_infos,
+                           std::back_inserter(table_ids),
+                           [](auto& info) { return info.id; });
+    // TODO: avoid forcing a new active segment if current active segment does
+    //       not contain any dirty mutation to the specified tables
```

Comment:
I'm not sure that's true in all cases, if a memtable is mutated over a long period of time some mutations in it may be in non-active CL segments as well.

I don't understand why call `flush_tables_on_all_shards` here while soon we call `flush_all_tables`.
I think the minimum fix is to just call `flush_all_tables`.
Later, there could be an optimization that will selectively flush all (mem)tables that hold a reference to non-active CL segments (after forcing a new active one).
This subset may be different than `_table_infos` for this keyspace.

Reply:
> I'm not sure that's true in all cases, if a memtable is mutated over a long period of time some mutations in it may be in non-active CL segments as well.

why is this related to memtable? mutations should be persisted into commit log or sstable. what  "gets back" is the ones in commitlog.

> 
> I don't understand why call `flush_tables_on_all_shards` here while soon we call `flush_all_tables`. 

i removed the `flush_all_tables()` call in the 2nd revision.

> I think the minimum fix is to just call `flush_all_tables`.

right. but we don't need to flush all of them. `flush_tables_on_all_shards()` is a low-hanging optimization.

> Later, there could be an optimization that will selectively flush all (mem)tables that hold a reference to non-active CL segments (after forcing a new active one). This subset may be different than `_table_infos` for this keyspace.

Reply:
> we don't need to flush all of them. 

right

> `flush_tables_on_all_shards()` is a low-hanging optimization

But it's not flushing the required subset of all tables.

`table::perform_cleanup_compaction` already flushes each table from the same set of tables as this call below.

Instead, we can start with flushing all tables, and proving it works with a test, then seek optimization by forcing new active commitlog segment and then flushing all tables pinning the non-active segments.

Reply:
> `table::perform_cleanup_compaction` already flushes each table from the same set of tables as this call below.
> 

@bhalevy i checked the code path of `table::perform_cleanup_compaction` all the way down to `compaction_task_executor::compact_sstables()` twice. but still cannot find the call site of `compaction_group::flush()`, which should be the place we flush memtables.

> Instead, we can start with flushing all tables, and proving it works with a test, then seek optimization by forcing new active commitlog segment and then flushing all tables pinning the non-active segments.

yeah, i agree. once i understand your point i will fall back to the safer approach. meanwhile, i will create a test in this week.

Reply:
> i checked the code path of table::perform_cleanup_compaction all the way down to compaction_task_executor::compact_sstables() twice. but still cannot find the call site of compaction_group::flush(), which should be the place we flush memtables.

Isn't that the first line of `table::perform_cleanup_compaction()`?:
https://github.com/scylladb/scylladb/blob/423234841e5c79caf2ef0d1afe5d018ea8f4afe2/replica/table.cc#L1563-L1565

`flush()` calls `co_await parallel_foreach_compaction_group(std::mem_fn(&compaction_group::flush));`